File: application/src/main/java/com/study/platform/domain/chat/application/SaveAndPublishChatMessageService.java
package com.study.platform.domain.chat.application;

import com.study.platform.domain.chat.dto.request.ChatMessageRequest;
import com.study.platform.domain.chat.model.ChatMessage;
import com.study.platform.domain.chat.model.ChatMessageRepository;
import com.study.platform.domain.chat.model.ChatPublisher;
import com.study.platform.domain.chat.usecase.SaveAndPublishChatMessageUseCase;
import com.study.platform.domain.team.model.StudyTeam;
import com.study.platform.domain.team.model.StudyTeamRepository;
import com.study.platform.domain.team.model.TeamMemberRepository;
import com.study.platform.domain.user.model.User;
import com.study.platform.domain.user.model.UserRepository;
import com.study.platform.global.exception.CustomException;
import com.study.platform.global.exception.ErrorCode;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Service
@RequiredArgsConstructor
public class SaveAndPublishChatMessageService implements SaveAndPublishChatMessageUseCase {

    private final ChatMessageRepository chatMessageRepository;
    private final StudyTeamRepository studyTeamRepository;
    private final TeamMemberRepository teamMemberRepository;
    private final UserRepository userRepository;
    private final ChatPublisher chatPublisher;

    @Override
    @Transactional
    public void execute(UUID userId, UUID teamId, ChatMessageRequest request) {
        if (!teamMemberRepository.existsByTeamIdAndUserId(teamId, userId)) {
            throw new CustomException(ErrorCode.NOT_TEAM_MEMBER);
        }

        StudyTeam team = studyTeamRepository.findById(teamId)
                .orElseThrow(() -> new CustomException(ErrorCode.TEAM_NOT_FOUND));
        User sender = userRepository.findById(userId)
                .orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND));

        ChatMessage message = chatMessageRepository.save(ChatMessage.create(team, sender, request.content()));
        chatPublisher.publish(message);
    }
}

File: application/src/main/kotlin/com/study/platform/domain/chat/application/SaveAndPublishChatMessageService.kt
package com.study.platform.domain.chat.application

import com.study.platform.domain.chat.dto.request.ChatMessageRequest
import com.study.platform.domain.chat.model.ChatMessage
import com.study.platform.domain.chat.model.ChatMessageRepository
import com.study.platform.domain.chat.model.ChatPublisher
import com.study.platform.domain.chat.usecase.SaveAndPublishChatMessageUseCase
import com.study.platform.domain.team.model.StudyTeamRepository
import com.study.platform.domain.team.model.TeamMemberRepository
import com.study.platform.domain.user.model.UserRepository
import com.study.platform.global.exception.CustomException
import com.study.platform.global.exception.ErrorCode
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.transaction.support.TransactionSynchronization
import org.springframework.transaction.support.TransactionSynchronizationManager
import java.util.UUID

@Service
class SaveAndPublishChatMessageService(
    private val chatMessageRepository: ChatMessageRepository,
    private val studyTeamRepository: StudyTeamRepository,
    private val teamMemberRepository: TeamMemberRepository,
    private val userRepository: UserRepository,
    private val chatPublisher: ChatPublisher
) : SaveAndPublishChatMessageUseCase {

    @Transactional
    override fun execute(userId: UUID, teamId: UUID, request: ChatMessageRequest) {
        if (!teamMemberRepository.existsByTeamIdAndUserId(teamId, userId)) {
            throw CustomException(ErrorCode.NOT_TEAM_MEMBER)
        }
        val team = studyTeamRepository.findById(teamId)
            ?: throw CustomException(ErrorCode.TEAM_NOT_FOUND)
        val sender = userRepository.findById(userId)
            ?: throw CustomException(ErrorCode.USER_NOT_FOUND)
        val message = chatMessageRepository.save(ChatMessage.create(team, sender, request.content))
        if (TransactionSynchronizationManager.isActualTransactionActive()) {
            TransactionSynchronizationManager.registerSynchronization(object : TransactionSynchronization {
                override fun afterCommit() {
                    chatPublisher.publish(message)
                }
            })
        } else {
            chatPublisher.publish(message)
        }
    }
}
